### ISSUE TITLE
The priority inversion when using the completion block.

### ISSUE THREAD

**Issue by Xezun** [SDWebImage/SDWebImage#3075]
### New Issue Checklist

* [ - ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ - ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ - ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.0.0 / 4.4.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 11 / Xcode 10
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

1. Create a new project and copy the codes bellow. (you also need install the SDWebImage)
2. Run the project, and the image will load successfully, and the viewController's backgroundColor is blue.
3. Stop and re-run the project, the viewController's backgroundColor may be red sometimes.

It not always happens, but quite frequently.

Here is the code:

```Objective-C
// in UIViewController's viewDidLoad method.

SDAnimatedImageView *imageView1 = [[SDAnimatedImageView alloc] initWithFrame:CGRectMake(80, 20, 200, 100)];
[self.view addSubview:imageView1];

NSURL *url = [NSURL URLWithString:@"https://support.apple.com/content/dam/edam/applecare/images/en_US/psp/featured-section-promo-siri-icon_2x.png"];
NSLog(@"0.1: %@", url);
[imageView1 sd_setImageWithURL:url placeholderImage:nil options:(SDWebImageAllowInvalidSSLCertificates) completed:^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
    self.view.backgroundColor = UIColor.redColor;
    NSLog(@"1.1: %@, %ld, %@", image, error.code, imageURL);
}];

dispatch_async(dispatch_get_main_queue(), ^{
    NSLog(@"0.2: %@", url);
    [imageView1 sd_setImageWithURL:url placeholderImage:nil options:(SDWebImageAllowInvalidSSLCertificates) completed:^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
        self.view.backgroundColor = UIColor.blueColor;
        NSLog(@"1.2: %@, %ld, %@", image, error.code, imageURL);
    }];
});
```

The console logs:

```log
2020-08-26 14:52:06.185480+0800 SDWebImageTest[5993:2461588] 0.1: https://support.apple.com/content/dam/edam/applecare/images/en_US/psp/featured-section-promo-siri-icon_2x.png
2020-08-26 14:52:06.276771+0800 SDWebImageTest[5993:2461588] 0.2: https://support.apple.com/content/dam/edam/applecare/images/en_US/psp/featured-section-promo-siri-icon_2x.png
2020-08-26 14:52:06.277425+0800 SDWebImageTest[5993:2461588] 1.2: <SDAnimatedImage:0x280b9c0b0 anonymous {800, 400}>, 0, https://support.apple.com/content/dam/edam/applecare/images/en_US/psp/featured-section-promo-siri-icon_2x.png
2020-08-26 14:52:06.277610+0800 SDWebImageTest[5993:2461588] 1.1: (null), 2002, https://support.apple.com/content/dam/edam/applecare/images/en_US/psp/featured-section-promo-siri-icon_2x.png
```

**Comment by dreampiggy** [comment]
Why you think these two call for `sd_setImageWithURL:` have any gurantee about the callback order ?
No any documentation or behavior talked about this.

**Comment by xezun** [comment]
> Why you think these two call for `sd_setImageWithURL:` have any gurantee about the callback order ?
> No any documentation or behavior talked about this.

Because sometimes we need re-layout the views if the image could not display correctly.

**Comment by xezun** [comment]
> Why you think these two call for `sd_setImageWithURL:` have any gurantee about the callback order ?
> No any documentation or behavior talked about this.

I think it's a bug rather than a feature, the SD should send cancel message before a new loading operation.

**Comment by dreampiggy** [comment]
cancel message need timing to wait.

For example, if the download process already started, you `cancel previous load` means,  cancel a NSURLSessionTask, which means, send TCP client reset, and wait for response. This may takes up to nearly 1 or more seconds.

So, which means, you next URL query (the code `NSLog(@"0.2: %@", url);`). Have to delay 1 or more seconds.

**Comment by dreampiggy** [comment]
Your issue sounds like a new feature request (behavior changes). Previouslly we never have any gurantee about the order. This need some changes inside the SDWebImageManager && SDWebImageDownloader communication. Marked into future PR.

**Comment by dreampiggy** [comment]
Still need a solution :(
